point set conversion

thewtex opened the pull request:
- **ENH: Bump version to 0.4.0**
- **ENH: Initial points_layer_from_point_set, point_set_from_points_layer**
- **DOC: Describe point set conversion, expand on image conversion**
- **DOC: Add support for applying points layer rotate**

thewtex committed:
ENH: Bump version to 0.4.0

thewtex committed:
ENH: Initial points_layer_from_point_set, point_set_from_points_layer

thewtex committed:
DOC: Describe point set conversion, expand on image conversion

thewtex committed:
DOC: Add support for applying points layer rotate

thewtex commented:
Closes #2

N-Dekker commented on README.md:
```diff
@@ -6,32 +6,223 @@
+# The ITK image will have:
+# - spacing in ITK order (reversed from napari)
+# - origin set from translate
+# - direction matrix from rotate
```
Somewhat more specific:

```py
# - spacing:  coordinates in ITK order, so reversed from napari `scale`: [1.5, 1.5, 2.0]
# - origin: coordinates in ITK order, so reversed from napari `translate`: [30.0, 20.0, 10.0]
# - direction: from napari `rotate`. In this case unspecified, so just the identity matrix.

```
Right?

thewtex replied:
@N-Dekker yes, thanks, Niels! Added.

thewtex committed:
DOC: More detailed description of image information correspondance

N-Dekker commented on README.md:
```diff
@@ -6,32 +6,223 @@
+# The layer will have:
+# - image_layer.scale = image spacing (in reverse order for NumPy)
+# - image_layer.translate = image origin
+# - image_layer.rotate = image direction matrix
```
`translate` is also in reverse order, but `rotate` is _not_ in reverse order, right? Or are they all in reverse order?

thewtex replied:
I think they are all in reverse order (?) but I have not see documentation on this if it exists. I will add the note to all three.

N-Dekker commented on README.md:
```diff
@@ -6,32 +6,223 @@
+# The layer will have:
+# - image_layer.scale = image spacing (in reverse order for NumPy)
+# - image_layer.translate = image origin (in reverse order for NumPy)
+# - image_layer.rotate = image direction matrix (in reverse order for NumPy)
```
`image_layer.rotate` is the _transpose_ of ITK's image direction matrix, right? (Rather than "in reverse order"). At least, that's what I understand from your https://github.com/InsightSoftwareConsortium/itk-napari-conversion/commit/7e48ef0d70df700851709dad37444880d636824d

thewtex replied:
There is the issue of the order elements of the columns / rows (which I am trying to indicate with "in reverse order", and which I assume is correct), and there is the issue of how the matrix is applied, including whether a transpose is needed. The latter is a separate issue, and I think it is harder to make an assumption, so I would not like to make assertions here ;-).

N-Dekker commented on README.md:
```diff
@@ -6,32 +6,223 @@
+# The ITK image will have:
+# - spacing: coordinates in ITK order, so reversed from napari `scale`: [1.5, 1.5, 2.0]
+# - origin: coordinates in ITK order, so reversed from napari `translate`: [30.0, 20.0, 10.0]
+# - direction: from napari `rotate`. In this case unspecified, so just the identity matrix.
```
Can you possibly extend the example here with a non-identity rotation matrix:

```py
viewer.add_image(
    data,
    scale=[2.0, 1.5, 1.5],  # anisotropic spacing
    translate=[10.0, 20.0, 30.0],
    rotate=[ ... ],
```

And then possibly mention in the comment (line 100 here) what the ITK `direction` then would be?

thewtex replied:
@N-Dekker :+1: added in 
cbdfcd2

thewtex committed:
DOC: Add rotation example for images

Documentation and test covering napari Image Layer rotate and the
itk.Image direction as a NumPy array an itk.Matrix.

N-Dekker commented on tests.py:
```diff
@@ -133,3 +133,167 @@ def check_angle(angle):
+
+    assert np.array_equal(data, itk.array_view_from_image(image))
+
+    # napari rotate is transposed compared to ITK direction
+    assert np.allclose(rotate, np.transpose(np.array(image["direction"])))
```
Cool, thanks Matt!!! So the napari `rotate` is just the transpose of the ITK "direction", right?

N-Dekker replied:
Locally I got:

```
Non-orthogonal slicing is being requested, but is not fully supported. Data is displayed without applying an out-of-slice rotation or shear component.
Traceback (most recent call last):
  File "D:\X\Src\M\MyPythonApp\MyPythonApp.py", line 41, in <module>
    test_image_from_image_layer_rotate_3d()
  File "D:\X\Src\M\MyPythonApp\MyPythonApp.py", line 27, in test_image_from_image_layer_rotate_3d
    assert np.allclose(rotate, np.transpose(np.array(image["direction"])))
AssertionError
```

Could be my fault, still having another look  🤷

N-Dekker replied:
By the way, the `assert np.allclose(rotate, np.transpose(np.array(image["direction"])))` looks somewhat redundant, because the next lines of code check `image["direction"]` as well. Especially:

    expected_direction = np.transpose(rotate)
    assert np.allclose(direction_array, expected_direction)

N-Dekker replied:
Sorry for the noise, I was running your new unit test on the old itk_napari_conversion version!!! That causes test assert failures, of course  😄

N-Dekker reviewed (approved):
Thanks Matt! Looks like this will be very useful for our elastix-napari plugin as well 🥇

N-Dekker commented on itk_napari_conversion.py:
```diff
@@ -55,3 +60,90 @@ def image_from_image_layer(image_layer):
+
+    # Set points
+    if len(data) > 0:
+        points = itk.vector_container_from_array(data.astype(np.float32).flatten())
```
@thewtex Do I understand correctly that the end-user may still need to reverse the `x` and `y` coordinate of each point, before passing it to ITK or ITKElastix?

thewtex replied:
@N-Dekker yes -- and maybe we should be doing the re-ordering in these functions, or provide an option?

ITK is x,y,z -- I assume napari is z,y,x (should be verified)
